src/cli/argsource.rs

```diff
@@ -0,0 +1,54 @@
+/// Abstracts over reading the current process environment variables as a
+/// zero-cost abstraction to support threaded in-process testing.
+use std::env;
+
+pub trait ArgSource {
+    type Args: Iterator<Item = String>;
+
+    fn args(&self) -> Self::Args;
+}
+
+/// Implements ArgSource with `std::env::args`
+pub struct OsArgs {}
+
+impl ArgSource for OsArgs {
+    type Args = env::Args;
+    fn args(&self) -> Self::Args {
+        env::args()
+    }
+}
+
+/// Helper for ArgSource over `Vec<String>`
+pub struct VecArgs {
+    v: Vec<String>,
+    i: usize,
+}
+
+impl From<&Vec<String>> for VecArgs {
+    fn from(source: &Vec<String>) -> Self {
+        let v = source.clone();
+        VecArgs { v, i: 0 }
+    }
+}
+
+impl Iterator for VecArgs {
```

Comment:
I'd have been tempted to just set the iterator type for VecArgs to `Box<dyn Iterator<Item=String>>` and then return `Box::new(self.v.iter().cloned())` fom the `fn args()` but this is okay.

Reply:
I tried things in that space. std::slice::Iterator<String> returns &String. Hmm, I can go back and noodle on that, or we can do as another pass to make it tighter. I think the contract is the key thing.